libs/multicall/src/utils/getMulticallContract.ts

```diff
@@ -0,0 +1,19 @@
+import { Multicall3, Multicall3Abi } from '@cowprotocol/abis'
+import { Contract } from '@ethersproject/contracts'
+import type { Web3Provider } from '@ethersproject/providers'
+
+import { MULTICALL_ADDRESS } from '../const'
+
+const multicallContractsCache = new Map<Web3Provider, Contract>()
+
+export function getMulticallContract(provider: Web3Provider): Multicall3 {
+  if (!multicallContractsCache.has(provider)) {
+    const multicall = new Contract(MULTICALL_ADDRESS, Multicall3Abi, provider) as Multicall3
+
+    multicallContractsCache.set(provider, multicall)
```

Comment:
but we map cache the contract per provider....

is this right?
Shouldn't be cached by network only? 

Also, if the provider object is recreated, so it will the contract of MULTICALL. Wouldn't this make a big memory leak (it will never cleanup the cache)

Reply:
>Shouldn't be cached by network only?

No, because initially are not connected to a wallet and you provider comes from `networkConnection`.
Then you connect a wallet without changing network and multicall should use the new provider.

>if the provider object is recreated
I've checked several times, provider recreates only when you connect a new wallet. So, you need to changes wallets thousands of times to feel any problems with memory.

Reply:
okokk